Fix sdl2-gfx, sdl2-image, sdl2-mixer, sdl2-net, and sdl2-ttf forcing debug lib first (SDL2d)

amaiorano opened the pull request:
When using sdl2-net on Windows, release builds end up linking against SDL2d.lib since it comes first in find_library. Removing all of this since it's not necessary. If using SDL2-net, it's reasonable to assume the client is already also linking against SDL2. Note that this was Windows-only, so for cross-platform builds, explicitly linking against SDL2 was already necessary.

Rastaban commented:
Could you also bump the CONTROL version number up to `2.0.1-6`?  This will force a rebuild on `vcpkg upgrade`

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5948) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5948) before we can accept your contribution.<br/><br/>:x: amaiorano [sign now](https://cla.opensource.microsoft.com/Microsoft/vcpkg?pullRequest=5948)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/vcpkg?pullRequest=5948) it.</sub>

Rastaban commented:
I am seeing a new build failure on `arm64-windows`, `x64-windows`, and `x86-windows` with this PR.  Can you look into them?

amaiorano commented:
> I am seeing a new build failure on `arm64-windows`, `x64-windows`, and `x86-windows` with this PR. Can you look into them?

I suppose you're referring to the Vcpkg-PR-Eager failure? It seems I'm not authorized to access it. It asks me to log in with a Microsoft account; but when I use the same one that I use to sign into Visual Studio, it tells me:

![image](https://user-images.githubusercontent.com/6893883/55629357-02bf4d00-5781-11e9-83b5-4050d9fe4168.png)

amaiorano commented:
I haven't been able to look at the tests, but I did take a look at the CMakeLists.txt files under ports/ for sdl2-gfx, sdl2-ttf, sdl2-image, and sdl2-mixer, and all of these do the same thing as sdl2-net. They all search for the SDL2 lib like so:

```
find_library(SDL_LIBRARY NAMES SDL2d SDL2)
```

In my case, I'm using sdl2-net, and what ends up happening is that because it searches for SDL2d before SDL2, it ends up finding it:

![image](https://user-images.githubusercontent.com/6893883/55630336-6e0a1e80-5783-11e9-9349-be2236c16c59.png)

As you can see, my Release build picks up SDL2d.lib. It also picks up SDL2.lib below, but that comes from my dependency on sdl2 directly.

I'm not sure what the right solution is. I tried something else: I modified the sdl2-net port CMakeLists.txt so that instead of ```find_library(SDL_LIBRARY NAMES SDL2d SDL2)```, it does ```find_package(SDL2 CONFIG REQUIRED)``` instead, and rather than add ```${SDL_LIBRARY_NAMES}``` to target_link_libraries, it now adds ```SDL2::SDL2-static```. So basically, using the modern config pattern to depend on SDL2 instead of the old find_library method. This works for me locally, and has the added benefit of only adding the dependency on SDL2.lib (or SDL2d.lib) once. I'll push this as a new commit, and we'll see if that passes tests.

amaiorano commented:
Looks like Vcpkg-PR-Eager has passed on this second commit. If you think this is good, I could apply this same fix to sdl2-gfx, sdl2-ttf, sdl2-image, and sdl2-mixer.

Rastaban commented:
This looks good to me.  Do you want to apply the other changes in this PR?

amaiorano commented:
> This looks good to me. Do you want to apply the other changes in this PR?

Alright, will do, and will rewrite history and update the PR name. Note that I don't use the other libs, so we'll have to rely on the tests, but really, if it works for one, it should work for the others.

amaiorano commented:
Btw, I didn't get a response about not being able to access Vcpkg-PR-Eager. Is there something I need to do?

Rastaban commented:
Vcpkg-PR-Eager is hosted on a Microsoft Internal Azure DevOps instance, its access is not controlled by the vcpkg team so we have been pasting the results into PRs.  We are working through the steps to get it moved to https://dev.azure.com/vcpkg/ where we will be able to open it up to the community.  This will hopefuly happen soon.

past-due commented:
@amaiorano: What about conditionally linking to `SDL2::SDL2` and `SDL2::SDL2-static` depending on the value of `VCPKG_LIBRARY_LINKAGE`?

```cmake
if(VCPKG_LIBRARY_LINKAGE STREQUAL "static")
   target_link_libraries(SDL2_net SDL2::SDL2-static)
else()
   target_link_libraries(SDL2_net SDL2::SDL2)
endif()
```

amaiorano commented:
> @amaiorano: What about conditionally linking to `SDL2::SDL2` and `SDL2::SDL2-static` depending on the value of `VCPKG_LIBRARY_LINKAGE`?

Ah, I thought vcpkg only created static libraries. I didn't know about VCPKG_LIBRARY_LINKAGE. I can certainly do as you propose. @Rastaban does that make sense to you as well?

amaiorano commented:
TBH, I don't even understand why SDL2 would require us to specify either SDL2-static or just SDL2 - can't it figure it out based on the CMake environment? Locally, I tried modifying my project to use SDL2::SDL2, despite only building static libs in vcpkg, and it worked...

past-due commented:
vcpkg has a wrapper-workaround (in the SDL2 port: see [`sdl2/vcpkg-cmake-wrapper.cmake`](https://github.com/Microsoft/vcpkg/blob/master/ports/sdl2/vcpkg-cmake-wrapper.cmake)) that defines one as the other if one is missing. (So things will at least compile.)

Ultimately, though, matching the linkage (and not relying on that wrapper) seems like a good goal.

amaiorano committed:
Fix sdl2-gfx, sdl2-image, sdl2-mixer, sdl2-net, and sdl2-ttf forcing debug lib first (SDL2d)

amaiorano commented:
Okay, I added support for both static and dynamic SDL2 libs. I've squashed all and force pushed, updated title, etc. Hopefully this passes tests. @past-due would you be able to test this locally as well?

past-due commented:
@amaiorano: I'm not sure that `VCPKG_LIBRARY_LINKAGE` is available in the CMakeLists.txt by default (I believe it's just defined in the portfile). See my edited comment above. Your changes will _work_, but I think they may always be hitting the `else()` case and then relying on the `sdl2/vcpkg-cmake-wrapper.cmake`.

@Rastaban: Thoughts on this (and the above)? Should we bother trying to configure the "appropriate" SDL2 target by linkage, or just pick one and rely on the `sdl2/vcpkg-cmake-wrapper.cmake`?

cenit commented:
What about using the official (but reversed in logic ;) ) `BUILD_SHARED_LIBS`, which is perfectly managed by vcpkg and correctly passed to the toolchain as true for shared builds?

Neumann-A commented:
will probably be solved by #5543

amaiorano commented:
Hey everyone,

So do we wait for #5543 to land, or do we proceed with this change?

Rastaban commented:
I lean towards waiting for the other PR to land.

cbezault committed:
Merge branch 'master' into fix-sdl2_net-forcing-sdl2-debug-lib

cbezault commented:
#5543 does not actually resolve the issue this PR is addressing. I'm still looking at the static/dynamic question more closely before merging.

cbezault committed:
Use BUILD_SHARED_LIBS in sdl2-gfx

cbezault committed:
Use BUILD_SHARED_LIBS in sdl2-image

cbezault committed:
Use BUILD_SHARED_LIBS in sdl2-mixer

cbezault committed:
Use BUILD_SHARED_LIBS in sdl2-net

cbezault committed:
Use BUILD_SHARED_LIBS in sdl2-ttf

cbezault committed:
Use BUILD_SHARED_LIBS correctly

cbezault committed:
Use BUILD_SHARED_LIBS correctly

cbezault committed:
Use BUILD_SHARED_LIBS correctly

cbezault committed:
Use BUILD_SHARED_LIBS correctly

cbezault committed:
Use BUILD_SHARED_LIBS correctly

ras0219-msft commented on ports/sdl2-net/CMakeLists.txt:
```diff
@@ -9,9 +9,9 @@ find_path(SDL_INCLUDE_DIR SDL2/SDL.h)
-else()
     set(SDL_LIBRARY SDL2::SDL2)
+else()
+    set(SDL_LIBRARY SDL2::SDL2-static)
```
A better option here is probably to use
```
if(TARGET SDL2::SDL2)
    set(SDL_LIBRARY SDL2::SDL2)
else()
    set(SDL_LIBRARY SDL2::SDL2-static)
endif()
```

It is not a guarantee that sdl2 was built in the same way that sdl2-net is being built -- it seems quite reasonable to dynamic link sdl2 but static link sdl2-net.

ras0219-msft replied:
~Actually, an even better option would be to add a `vcpkg-cmake-wrapper.cmake` file around SDL2 which creates one target as an alias to the other. Then all the downstream libraries would not need to play this boilerplate game.~

Wait, SDL2 already has a wrapper that is dealing with this[1]. Are you sure this is needed instead of just always linking `SDL2::SDL2`?
[1] https://github.com/Microsoft/vcpkg/blob/9db65c78914407b952f520e9cd09a7c2e27e4a96/ports/sdl2/vcpkg-cmake-wrapper.cmake

cbezault committed:
Always just use SDL2::SDL2

cbezault committed:
Remove SDL2_LIBRARY reference

cbezault committed:
Fix typo

amaiorano commented:
Thanks for wrapping this up and merging!
